Better structure the output of the list playbook

lhuard1A opened the pull request:
The list playbook listed the IPs of the VMs without logging their role like:

    TASK: [debug ] ************************************************************
    ok: [10.64.109.37] => {
        "msg": "public:10.64.109.37 private:192.168.165.5"
    }
    ok: [10.64.109.47] => {
        "msg": "public:10.64.109.47 private:192.168.165.6"
    }
    ok: [10.64.109.36] => {
        "msg": "public:10.64.109.36 private:192.168.165.4"
    }
    ok: [10.64.109.215] => {
        "msg": "public:10.64.109.215 private:192.168.165.2"
    }

The list playbook now prints the information in a more structured way with
a list of masters, a list of nodes and the subtype of the nodes like:

    TASK: [debug ] ************************************************************
    ok: [localhost] => {
        "msg": {
            "lenaicnewlist": {
                "master": [
                    {
                        "name": "10.64.109.215",
                        "private IP": "192.168.165.2",
                        "public IP": "10.64.109.215",
                        "subtype": "default"
                    }
                ],
                "node": [
                    {
                        "name": "10.64.109.47",
                        "private IP": "192.168.165.6",
                        "public IP": "10.64.109.47",
                        "subtype": "compute"
                    },
                    {
                        "name": "10.64.109.37",
                        "private IP": "192.168.165.5",
                        "public IP": "10.64.109.37",
                        "subtype": "compute"
                    },
                    {
                        "name": "10.64.109.36",
                        "private IP": "192.168.165.4",
                        "public IP": "10.64.109.36",
                        "subtype": "infra"
                    }
                ]
            }
        }
    }

lhuard1A commented:
I modified only the `openstack` and `libvirt` providers because I don’t have any AWS and GCE accounts to test them.

detiber commented:
Other than the test errors, lgtm.

Test errors:
```
GitHub pull request #461 of commit 8e805994b80a0d45659953db63d91b002711ef51, no merge conflicts.
Setting status of 8e805994b80a0d45659953db63d91b002711ef51 to PENDING with url https://ci.openshift.redhat.com/jenkins/job/openshift-ansible/667/ and message: 'Build started sha1 is merged.'
Building in workspace /var/lib/jenkins/jobs/openshift-ansible/workspace
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url git@github.com:openshift/openshift-ansible.git # timeout=10
Fetching upstream changes from git@github.com:openshift/openshift-ansible.git
 > git --version # timeout=10
 > git fetch --tags --progress git@github.com:openshift/openshift-ansible.git +refs/head/*:refs/remotes/origin/* +refs/pull/*:refs/remotes/origin/pr/*
 > git rev-parse refs/remotes/origin/pr/461/merge^{commit} # timeout=10
 > git rev-parse refs/remotes/origin/origin/pr/461/merge^{commit} # timeout=10
Checking out Revision afc2d1655e060c7df2365c7c533c2483dd7960d2 (refs/remotes/origin/pr/461/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f afc2d1655e060c7df2365c7c533c2483dd7960d2
First time build. Skipping changelog.
[workspace] $ /bin/sh -xe /tmp/hudson6124940148823935787.sh
[workspace] $ /bin/sh -xe /tmp/hudson7967472025184306575.sh
+ git/yaml_validation.rb
++ /usr/bin/git rev-parse 'HEAD^'
+ echo 'cd989f62709db9f858205420869b9a665cc609a1 afc2d1655e060c7df2365c7c533c2483dd7960d2 master'
++++++ Received:  /tmp/afc2d1655e060c7df2365c7c533c2483dd7960d220150813-11272-1v365to/playbooks/libvirt/openshift-cluster/list.yml
++++++ Received:  /tmp/afc2d1655e060c7df2365c7c533c2483dd7960d220150813-11272-1v365to/playbooks/libvirt/openshift-cluster/tasks/launch_instances.yml
++++++ Received:  /tmp/afc2d1655e060c7df2365c7c533c2483dd7960d220150813-11272-1v365to/playbooks/libvirt/openshift-cluster/templates/domain.xml
++++++ Received:  /tmp/afc2d1655e060c7df2365c7c533c2483dd7960d220150813-11272-1v365to/playbooks/openstack/openshift-cluster/list.yml
++++++ Received:  /tmp/afc2d1655e060c7df2365c7c533c2483dd7960d220150813-11272-1v365to/filter_plugins/oo_filters.py
[workspace] $ /bin/sh -xe /tmp/hudson3413897543416179487.sh
+ source /var/lib/jenkins/python27/bin/activate
++ /usr/bin/git rev-parse 'HEAD^'
+ ./git/pylint.sh cd989f62709db9f858205420869b9a665cc609a1 afc2d1655e060c7df2365c7c533c2483dd7960d2 master
Testing files:  filter_plugins/oo_filters.py
************* Module oo_filters
C:133, 0: Trailing whitespace (trailing-whitespace)
C:142, 0: No space allowed after bracket
        return out_joiner.join([ in_joiner.join([k, v]) for k, v in data.items() ])
                               ^ (bad-whitespace)
C:142, 0: No space allowed before bracket
        return out_joiner.join([ in_joiner.join([k, v]) for k, v in data.items() ])
                                                                                 ^ (bad-whitespace)
Build step 'Execute shell' marked build as failure
Setting status of 8e805994b80a0d45659953db63d91b002711ef51 to FAILURE with url https://ci.openshift.redhat.com/jenkins/job/openshift-ansible/667/ and message: 'Build finished. No test results found.'
Finished: FAILURE
```

lhuard1A commented:
@detiber : It seems that those pylint errors are coming from `master` and not from my PR. (Those lines are untouched in that PR).
In order to keep git history as clean as possible, I created a dedicated PR to fix those errors: #473 
Once #473 is merged, we could re-launched Jenkins here.

jaryn commented:
Can one of the admins verify this patch?

abutcher commented:
I like this a lot but it doesn't fail nicely with an empty or incorrect `cluster_id`. I think we have some examples where we use an empty string for `cluster_id`.

```
$ ./bin/cluster list libvirt ''                            

...

PLAY [List Hosts] ************************************************************* 

TASK: [debug ] **************************************************************** 
fatal: [localhost] => Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 582, in _executor
    exec_rc = self._executor_internal(host, new_stdin)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 785, in _executor_internal
    return self._executor_internal_inner(host, self.module_name, self.module_args, inject, port, complex_args=complex_args)
  File "/usr/lib/python2.7/site-packages/ansible/runner/__init__.py", line 1009, in _executor_internal_inner
    complex_args = template.template(self.basedir, complex_args, inject, fail_on_undefined=self.error_on_undefined_vars)
  File "/usr/lib/python2.7/site-packages/ansible/utils/template.py", line 138, in template
    d[k] = template(basedir, v, templatevars, lookup_fatal, depth, expand_lists, convert_bare, fail_on_undefined, filter_fatal)
  File "/usr/lib/python2.7/site-packages/ansible/utils/template.py", line 122, in template
    varname = template_from_string(basedir, varname, templatevars, fail_on_undefined)
  File "/usr/lib/python2.7/site-packages/ansible/utils/template.py", line 371, in template_from_string
    res = jinja2.utils.concat(rf)
  File "<template>", line 12, in root
  File "/home/abutcher/rhat/openshift-ansible/filter_plugins/oo_filters.py", line 373, in oo_pretty_print_cluster
    env=_get_tag_value(host['group_names'], 'env'),
  File "/home/abutcher/rhat/openshift-ansible/filter_plugins/oo_filters.py", line 357, in _get_tag_value
    raise KeyError(key)
KeyError: 'env'
```

As opposed to

```
$ ./bin/cluster list libvirt ''           

PLAY [Generate oo_list_hosts group] ******************************************* 

TASK: [set_fact scratch_group=tag_env-{{ cluster_id }}] *********************** 
skipping: [localhost]

TASK: [set_fact scratch_group=all] ******************************************** 
ok: [localhost]

TASK: [add_host ] ************************************************************* 
skipping: [localhost]

PLAY [List Hosts] ************************************************************* 
skipping: no hosts matched

PLAY RECAP ******************************************************************** 
localhost                  : ok=1    changed=0    unreachable=0    failed=0   

```

AND

```
$ ./bin/cluster list libvirt 'thisiswrong'

...

PLAY [List Hosts] ************************************************************* 

TASK: [debug ] **************************************************************** 
fatal: [localhost] => Failed to template {{ hostvars | oo_select_keys(groups[scratch_group]) | oo_pretty_print_cluster }}: |failed expects first param is a list

FATAL: all hosts have already failed -- aborting
```

As opposed to

```
$ ./bin/cluster list libvirt 'thisiswrong'

PLAY [Generate oo_list_hosts group] ******************************************* 

TASK: [set_fact scratch_group=tag_env-{{ cluster_id }}] *********************** 
ok: [localhost]

TASK: [set_fact scratch_group=all] ******************************************** 
skipping: [localhost]

TASK: [add_host ] ************************************************************* 
skipping: [localhost]

PLAY [List Hosts] ************************************************************* 
skipping: no hosts matched

PLAY RECAP ******************************************************************** 
localhost                  : ok=1    changed=0    unreachable=0    failed=0   

```

jaryn commented:
Can one of the admins verify this patch?

arilivigni commented:
Testing job has been triggered.

arilivigni commented:
Internal Jenkins error during triggering.
 No test results found.

lhuard1A commented:
I have just rebased that PR

@abutcher I have also fixed the two cases you reported.

Here is what we get if we provide a non-existent cluster-id: (an empty list)
```
$ bin/cluster list libvirt "wrong"

…
PLAY [List Hosts] ************************************************************* 

TASK: [debug ] **************************************************************** 
ok: [localhost] => {
    "msg": {}
}

PLAY RECAP ******************************************************************** 
localhost                  : ok=2    changed=0    unreachable=0    failed=0   

```

And here is what we get if we provide an empty cluster-id: (it lists all the clusters)
```
$ bin/cluster list libvirt ""

...
PLAY [List Hosts] ************************************************************* 

TASK: [debug ] **************************************************************** 
ok: [localhost] => {
    "msg": {
        "bar": {
            "master": [
                {
                    "name": "bar-master-8b902", 
                    "private IP": "192.168.55.189", 
                    "public IP": "192.168.55.189", 
                    "subtype": "default"
                }
            ], 
            "node": [
                {
                    "name": "bar-node-compute-97ee3", 
                    "private IP": "192.168.55.32", 
                    "public IP": "192.168.55.32", 
                    "subtype": "compute"
                }
            ]
        }, 
        "foo": {
            "master": [
                {
                    "name": "foo-master-fbd6d", 
                    "private IP": "192.168.55.13", 
                    "public IP": "192.168.55.13", 
                    "subtype": "default"
                }
            ], 
            "node": [
                {
                    "name": "foo-node-compute-1d3ef", 
                    "private IP": "192.168.55.220", 
                    "public IP": "192.168.55.220", 
                    "subtype": "compute"
                }
            ]
        }
    }
}

PLAY RECAP ******************************************************************** 
bar-master-8b902           : ok=1    changed=0    unreachable=0    failed=0   
bar-node-compute-97ee3     : ok=1    changed=0    unreachable=0    failed=0   
foo-master-fbd6d           : ok=1    changed=0    unreachable=0    failed=0   
foo-node-compute-1d3ef     : ok=1    changed=0    unreachable=0    failed=0   
localhost                  : ok=3    changed=0    unreachable=0    failed=0   
```

detiber commented:
@lhuard1A shouldn't this also break host types down by subtype as well? or potentially, only break down by subtype if more than one subtype exists for a host type?

I'm thinking something in the neighborhood of:
```
env1:
  etcd:
    default:
    - name: etcd1
    ...
  master:
    default:
    - name: master1
    ...
  node:
    compute:
    - node-compute1
    ...
    infra:
    - name: node-infra1
    ...
```

abutcher commented:
@lhuard1A Confirmed empty and non-existent `cluster_id` output looks good :+1:

lhuard1A commented:
@detiber : I updated the code so that nodes are now grouped by subtype. Here is what the full output looks like:

```
PLAY [List Hosts] ************************************************************* 

TASK: [debug ] **************************************************************** 
ok: [localhost] => {
    "msg": {
        "lenaic": {
            "master": {
                "default": [
                    {
                        "name": "lenaic-master-a3d2a", 
                        "private IP": "192.168.55.100", 
                        "public IP": "192.168.55.100"
                    }
                ]
            }, 
            "node": {
                "compute": [
                    {
                        "name": "lenaic-node-compute-8923b", 
                        "private IP": "192.168.55.201", 
                        "public IP": "192.168.55.201"
                    }, 
                    {
                        "name": "lenaic-node-compute-ed1e9", 
                        "private IP": "192.168.55.5", 
                        "public IP": "192.168.55.5"
                    }
                ], 
                "infra": [
                    {
                        "name": "lenaic-node-infra-8a236", 
                        "private IP": "192.168.55.211", 
                        "public IP": "192.168.55.211"
                    }
                ]
            }
        }
    }
}
```

detiber commented:
@lhuard1A looks great, thanks.

@twiest @wshearn  lgtm

twiest commented:
merging based on @detiber's comment.
